Fix primary index not being used when a redundant equals exists

josh-hildred opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix an issue where when a redundant `= 1` or `= 0` is added after a boolean expression involving the primary key, the primary index is not used. For example, both `SELECT * FROM <table> WHERE <primary-key> IN (<value>) = 1` and `SELECT * FROM <table> WHERE <primary-key> NOT IN (<value>) = 0` will both perform a full table scan, when the primary index can be used.


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ] <!---ci_include_fast--> Fast test
- [ ] <!---ci_include_integration--> Integration Tests
- [ ] <!---ci_include_stateless--> Stateless tests
- [ ] <!---ci_include_stateful--> Stateful tests
- [ ] <!---ci_include_unit--> Unit tests
- [ ] <!---ci_include_performance--> Performance tests
- [ ] <!---ci_include_asan--> All with ASAN
- [ ] <!---ci_include_tsan--> All with TSAN
- [ ] <!---ci_include_analyzer--> All with Analyzer
- [ ] <!---ci_include_KEYWORD--> Add your option here

#### Exclude tests:
- [ ] <!---ci_exclude_fast--> Fast test
- [ ] <!---ci_exclude_integration--> Integration Tests
- [ ] <!---ci_exclude_stateless--> Stateless tests
- [ ] <!---ci_exclude_stateful--> Stateful tests
- [ ] <!---ci_exclude_performance--> Performance tests
- [ ] <!---ci_exclude_asan--> All with ASAN
- [ ] <!---ci_exclude_tsan--> All with TSAN
- [ ] <!---ci_exclude_msan--> All with MSAN
- [ ] <!---ci_exclude_ubsan--> All with UBSAN
- [ ] <!---ci_exclude_coverage--> All with Coverage
- [ ] <!---ci_exclude_aarch64--> All with Aarch64
- [ ] <!---ci_exclude_KEYWORD--> Add your option here

#### Extra options:
- [ ] <!---do_not_test--> do not test (only style check)
- [ ] <!---no_merge_commit--> disable merge-commit (no merge from master before tests)
- [ ] <!---no_ci_cache--> disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ] <!---batch_0--> 1
- [ ] <!---batch_1--> 2
- [ ] <!---batch_2--> 3
- [ ] <!---batch_3--> 4

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #62142 from ClibMouse/ClickHouse:fixes_in_with_pk_ft_scan -->
*This is an automated comment for commit 8ac218bd849d3ced26e987de3b7178b765410242 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62142/8ac218bd849d3ced26e987de3b7178b765410242/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

josh-hildred committed:
Add an optimization that removes redundant equality
checks on boolean functions. This fixes a bug in
which the primary index is not used for queries like
SELECT * FROM <table> WHERE <pk> in (<n>) = 1

josh-hildred commented:
Test failures appear unrelated

vdimir reviewed (commented):
Thanks for the PR!

The idea is good and clear.
 I'm thinking of wouldn't it be better to implement such optimization before KeyCondition is applied, for example in https://github.com/ClickHouse/ClickHouse/blob/master/src/Analyzer/Passes/LogicalExpressionOptimizerPass.h that is executed on top of Query Tree.

We do have similar optimizations there, so maybe we can just rewrite `equals` that accept a function returning bool and `0/1` constant to the expression itself. So, futher query execution steps will receive expression without ` ... == 1` part.

josh-hildred commented:
Thanks for the review @vdimir! I moved all the functionality to `LogicalExpressionOptimizerVisitor` in https://github.com/ClickHouse/ClickHouse/blob/master/src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp. Please let me know if you have any more comments!

josh-hildred committed:
Fix style

josh-hildred committed:
Fix an isssue with constants being wrapped in nullables

josh-hildred committed:
Update tests to set allow_experimental_analyzer

josh-hildred committed:
Add additional tests for queries with toLowCardinality and toNullable

vdimir commented on src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp:
```diff
@@ -552,6 +582,67 @@ class LogicalExpressionOptimizerVisitor : public InDepthQueryTreeVisitorWithCont
+        auto & function_node = node->as<FunctionNode &>();
+        assert(function_node.getFunctionName() == "equals");
+
+        bool lhs_const;
```
nitpick: I'd rewrite code below a bit, to declare variables closer to first usages, like this (I haven't checked the code below, so feel free to change it or just discard suggestion):

```cpp
        const auto function_arguments = function_node.getArguments().getNodes();
        if (function_arguments.size() != 2)
            return;

        const auto & lhs = function_arguments[0];
        const auto & rhs = function_arguments[1];

        UInt64 constant_value;
        bool is_lhs_const;
        if (const auto * lhs_constant = lhs->as<ConstantNode>())
        {
            if (!lhs_constant->getValue().tryGet<UInt64>(constant_value) || constant_value > 1)
                return;
            is_lhs_const = true;
        }
        else if (const auto * rhs_constant = rhs->as<ConstantNode>())
        {
            if (!rhs_constant->getValue().tryGet<UInt64>(constant_value) || constant_value > 1)
                return;
            is_lhs_const = false;
        }
        else
            return;

        bool need_invert = (constant_value == 0);

        const FunctionNode * child_function = is_lhs_const ? rhs->as<FunctionNode>() : lhs->as<FunctionNode>();

        if (!child_function || !isBooleanFunction(child_function->getFunctionName()))
            return;
```

vdimir reviewed (commented):
Thanks, looks really good.
The main comment is about re-resolving, I assume it's better to add `CAST` if type is changed, check it out, please

vdimir commented on src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp:
```diff
@@ -552,6 +582,67 @@ class LogicalExpressionOptimizerVisitor : public InDepthQueryTreeVisitorWithCont
+            return;
+
+        if (removeLowCardinality(constant->getResultType())->isNullable())
+            need_rerun_resolve = true;
```
Oh, I see. Maybe we can not to consider this case for now and support only `UInt8` constant, because something like `= toNullable(1)` should be pretty rear case, wdyt?

josh-hildred committed:
Address review comments

josh-hildred committed:
Small code cleanup

vdimir reviewed (approved):
Awesome work on this PR! Thanks for your contribution.

josh-hildred commented:
Thank you again for the review @vdimir!
